source/common/upstream/health_checker_impl.cc

```diff
@@ -586,5 +595,297 @@ RedisHealthCheckerImpl::HealthCheckRequest::HealthCheckRequest() {
+    // received. Healthcheck error has already been reported by onResetStream() callback.
+    return;
+  }
+
+  grpc::health::v1::HealthCheckRequest request;
+  if (parent_.service_name_.valid())
+    request.set_service(parent_.service_name_.value());
+
+  request_encoder_->encodeData(*Grpc::Common::serializeBody(request), true);
+}
+
+void GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::onResetStream(Http::StreamResetReason) {
+  bool expect_reset = expect_reset_;
+  resetState();
+
+  if (expect_reset)
+    return;
+
+  ENVOY_CONN_LOG(debug, "connection/stream error health_flags={}", *client_,
+                 HostUtility::healthFlagsToString(*host_));
+
+  // TODO(baranov1ch): according to all HTTP standards, we should check if reason is one of
+  // Http::StreamResetReason::RemoteRefusedStreamReset (which may mean GOAWAY),
+  // Http::StreamResetReason::RemoteReset or Http::StreamResetReason::ConnectionTermination (both
+  // mean connection close), check if connection is not fresh (was used for at least 1 request)
+  // and silently retry request on the fresh connection. This is also true for HTTP/1.1 healthcheck.
+  handleFailure(FailureType::Network);
+
+  if (client_) {
+    // Force connection close on healthcheck errors and timeouts.
+    // Rationale: healthcheck error may mean cluster host is going down gracefully. Technically (
+    // for google-written gRPC server libs), it means that host sends us GOAWAY and PING, waiting
+    // PING ACK to initiate real connection close. But nghttp2 does not respond to pings received
+    // after GOAWAY, and envoy does not handle remote-initiated GOAWAYs yet
+    // (https://github.com/nghttp2/nghttp2/issues/1103).
+    // Altogether, this results in us holding the connection, preventing gRPC server graceful
+    // shutdown until server close timeout expires (typically 1 minute). This is bad. It's better to
+    // drop connection in case of errors than to mess up backend graceful shutdown. Still, all
+    // successful checks will be performed on a single HTTP/2 connection.
+    client_->close();
+  }
+}
+
+bool GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::isHealthCheckSucceeded(
+    Grpc::Status::GrpcStatus grpc_status) const {
+  if (grpc_status != Grpc::Status::GrpcStatus::Ok)
+    return false;
+
+  if (!health_check_response_ ||
+      health_check_response_->status() != grpc::health::v1::HealthCheckResponse::SERVING) {
+    return false;
+  }
+
+  return true;
+}
+
+void GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::onRpcComplete(
+    Grpc::Status::GrpcStatus grpc_status, const std::string& grpc_message, bool end_stream) {
+  logHealthCheckStatus(grpc_status, grpc_message);
+  if (isHealthCheckSucceeded(grpc_status)) {
+    handleSuccess();
+  } else {
+    handleFailure(FailureType::Active);
+  }
+
+  if (end_stream) {
+    resetState();
+  } else {
+    // resetState() will be called by onResetStream().
+    expect_reset_ = true;
```

Comment:
This logic is a bit confusing to me when comparing with `Grpc::AsyncClient`. Can we use the same close local/remote as done there?

Reply:
here we don't use local/remote close, logic is simpler:
- 'local' is always closed, RPC is unary, everything is sent by `onInterval()`
- `end_stream` argument is passed to `onRpcComplete` instead of using a field indicating 'remote close'.
Here local/remote close logic is not necessary, since there are no clients trying to send/receive streams of grpc messages.
- besides local/remote close, there is `http_reset_` in `Grpc::AsyncClient` which is used to determine if `Http::AsyncClient::Stream::reset()` should be called to recycle http stream, or stream was already reset by http stack. `expect_reset_ ` has nearly the same semantics as `http_reset_`. But unlike `Grpc::AsyncClient`, which maps stream reset just to internal error, health checker distinguishes between active healthcheck errors (when service is in really bad mood)  and network-ish healthcheck errors produced by stream reset.

Reply:
Fair enough. I think it would be helpful to comment here on why we might have seen an RPC completion, without hitting the end-of-stream, and why this might lead to a following `onResetStream()` callback. Thanks.

Reply:
added some comments. PTAL if they are good enough

Reply:
Yeah, these are good, thanks. Maybe also explain how we ever get into the situation where we are in this branch, i.e. who called `onRpcComplete` with `!end_stream` as well.

Reply:
done